cirq/contrib/acquaintance/permutation.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Sequence, Tuple, TypeVar
+
+from cirq import abc
+from cirq.ops import (
+        CompositeGate, Gate, QubitId, OP_TREE, SWAP,
+        flatten_op_tree, GateOperation, TextDiagrammable,
+        gate_features)
+
+LogicalIndex = TypeVar('LogicalIndex', int, QubitId)
+LogicalGates = Dict[Tuple[LogicalIndex, ...], Gate]
+LogicalMappingKey = TypeVar('LogicalMappingKey', bound=QubitId)
+LogicalMapping = Dict[LogicalMappingKey, LogicalIndex]
+
+class PermutationGate(Gate, TextDiagrammable, metaclass=abc.ABCMeta):
+    """Permutation gate."""
```

Comment:
More detail. For example, mention that you can pick the type of swapping and the permutation. And talk about how this is used for rewriting qubit ids.